tests/test_dataset.py

```diff
@@ -189,16 +195,355 @@ def test_cant_join_different_categories(self):
         s1 = Dataset.from_iterable([], categories=['a', 'b'])
         s2 = Dataset.from_iterable([], categories=['b', 'a'])
 
-        with self.assertRaisesRegex(Exception, "different categories"):
+        with self.assertRaisesRegex(DatumaroError, "different categories"):
             Dataset.from_extractors(s1, s2)
 
     def test_can_join_datasets(self):
         s1 = Dataset.from_iterable([ DatasetItem(0), DatasetItem(1) ])
         s2 = Dataset.from_iterable([ DatasetItem(1), DatasetItem(2) ])
+        expected = Dataset.from_iterable([
+            DatasetItem(0), DatasetItem(1), DatasetItem(2)
+        ])
+
+        actual = Dataset.from_extractors(s1, s2)
+
+        compare_datasets(self, expected, actual)
+
+    def test_inplace_save_writes_only_updated_data(self):
+        with TestDir() as path:
+            # generate initial dataset
+            dataset = Dataset.from_iterable([
+                DatasetItem(1, subset='a'),
+                DatasetItem(2, subset='b'),
+                DatasetItem(3, subset='c'),
+            ])
+            dataset.save(path)
+            ts1_a = os.stat(osp.join(path, 'annotations', 'a.json')).st_mtime_ns
+            ts1_b = os.stat(osp.join(path, 'annotations', 'b.json')).st_mtime_ns
+            ts1_c = osp.isfile(osp.join(path, 'annotations', 'c.json'))
+
+            dataset = Dataset.load(path)
+            dataset.put(DatasetItem(2, subset='a'))
+            dataset.remove(3, 'c')
+            time.sleep(1)
```

Comment:
Why do we need sleep here?

Reply:
In the test there are 3 subset files generated - a, b and c.
a should be updated, b should be kept the same, c should be removed.

There is a comparison between file modification times. Without the sleep the test was passing too fast, so that the comparison failed. The other way to implement the check can be through comparison of file hashes or something more clever.